Issue#104 Add label filtering

sergiughf opened the pull request:
### Resolves: https://github.com/telia-oss/github-pr-resource/issues/104

## Description

This PR adds the possibility to only fetch versions that have one of the specificied labels applied on the pull request. 

### What has been done

- Added `labels` field to `source` of type `[]string`
- When checking for new versions we filter out any commits from PRs that don't have one of the specified labels attached. If no labels specified on the source, there will be no filtering out of any commits. 
- ~When getting the resource we attached in the Metadata section the `labels` info.~
- Updates the README to reflect newly added feature.  


### Examples

New source config:
```yaml
resources:
  - name: pull-request
    type: pull-request
    icon: source-branch
    check_every: 6h
    source:
      repository: sergiughf/github-pr-resource
      access_token: ((secrets.github_token))
      labels: ["potato", "steak"]

```

### What is missing

~- I've noticed the E2E tests rely on an archived repository. Unfortunately I cannot add any new tests as I don't have the possibility of creating a PR with specific labels in there. I would kindly ask for your help on this one :)~

sergiughf commented on github.go:
```diff
@@ -97,57 +97,79 @@ func NewGithubClient(s *Source) (*GithubClient, error) {
 		}
+
 		for _, p := range query.Repository.PullRequests.Edges {
-			for _, c := range p.Node.Commits.Edges {
```
https://github.com/telia-oss/github-pr-resource/pull/132/files#diff-c411124e4d43a847f2f6360d74fba17eR138 we hard code to only get the last commit so I removed the loop. Still doesn't feel pretty so if you have any suggestions to improve this part they are more than welcome!

itsdalmo replied:
Your solution should work, but ideally I'd like to avoid changing the way this loop works since it is tasked with the only thing where this resource really needs to be consistent 😅 

Since labels are set per PR and not commit, and should be the same for every `&PullRequest{}` (only the `Tip` would differ if for some reason we got multiple commits back), you can just create a list of labels prior to entering the loop over commit edges:

```go
		var response []*PullRequest
	for {
		if err := m.V4.Query(context.TODO(), &query, vars); err != nil {
			return nil, err
		}
		for _, p := range query.Repository.PullRequests.Edges {
			labels := make([]LabelObject, len(p.Node.Labels.Edges))

			for _, l := range p.Node.Labels.Edges {
				labels = append(labels, l.Node.LabelObject)
			}

			for _, c := range p.Node.Commits.Edges {
				response = append(response, &PullRequest{
					PullRequestObject: p.Node.PullRequestObject,
					Tip:               c.Node.Commit,
					Labels:            labels,
				})
			}
		}
		if !query.Repository.PullRequests.PageInfo.HasNextPage {
			break
		}
		vars["prCursor"] = query.Repository.PullRequests.PageInfo.EndCursor
	}
```

Should work no?

sergiughf replied:
Yep definitely :) Thank you!

jmccann commented:
I did compile and try this myself.  I haven't tried the filtering piece yet but am in need of list of labels for a PR being provided and this did it wonderfully!  Thanks!

itsdalmo commented:
Hi @sergiughf - first of all, thank you for the PR and sorry I have been slow to review it - had to think a bit about where to go with filters (ref #58) first. With that out of the way, I'll try to get a review put together for you tomorrow - hope that is ok 😅

sergiughf commented:
Hey @itsdalmo no worries :)

itsdalmo commented on github.go:
```diff
@@ -273,6 +295,13 @@ func (m *GithubClient) GetPullRequest(prNumber, commitRef string) (*PullRequest,
 						}
 					}
 				} `graphql:"commits(last:$commitsLast)"`
+				Labels struct {
```
Hm, are we certain we need to fetch the label information again for `get` if we are not really doing anything with it?

sergiughf replied:
Please correct me if I am wrong but I assumed we could only populate the metadata section after we do a `get`. 

However based on the comment below I decided to remove it completely from the metadata so we don't end up doing this extra label call.

itsdalmo commented on in.go:
```diff
@@ -73,6 +73,16 @@ func Get(request GetRequest, github Github, git Git, outputDir string) (*GetResp
+		labels += pull.Labels[i].Name
+
+		if i < len(pull.Labels)-1 {
+			labels += ", "
```
You should use `strings.Join` here instead of rolling your own. 

But, I'm wondering if we should:
1. Not emit the metadata to be displayed by Concourse since it can be an arbitrary number. Right now you have limited it to 10 labels, but if Github allows it to be more, then I think it's more important to make sure the filter functionality is consistent (by fetching up to 100 labels) rather than being able to display it in the UI.
2. Fetch it and only write it to a file in `.git/resource`.
3. Not fetch the labels on `get` at all, since it's not being used by the resource outside of `check`.

I'm leaning towards number 2 or 3 here.

sergiughf replied:
> You should use strings.Join here instead of rolling your own.

:) Yeah went too old school. 

As for the other points, the only reason why I added the labels in the metadata is because I thought it might be useful for some users to see all of the labels being applied on the PR like so:

![image](https://user-images.githubusercontent.com/28300158/61381541-78bc6f80-a8ab-11e9-82e9-e10a40896bc7.png)

But I totally agree with you that the actual label filtering logic in the `check` is the most important part. We already know exactly what labels we will expect.

So I'm totally fine with not fetching the labels at all on `get`.

itsdalmo commented on models.go:
```diff
@@ -110,3 +112,12 @@ type CommitObject struct {
+
+// LabelObject represents the GraphQL label node.
+// https://developer.github.com/v4/object/label
+type LabelObject struct {
```
As far as I can tell, the only value we are actually using from the `LabelObject` is `Name`. We should not fetch the other values if that is the case, since it just makes the operation more expensive 🤷‍♂

sergiughf replied:
Removed all the others and only left `Name` :+1:

sergiughf commented on github.go:
```diff
@@ -97,57 +97,79 @@ func NewGithubClient(s *Source) (*GithubClient, error) {
+			"prStates":        []githubv4.PullRequestState{githubv4.PullRequestStateOpen},
+			"prCursor":        (*githubv4.String)(nil),
+			"commitsLast":     githubv4.Int(1),
+			"labelsFirst":     githubv4.Int(10),
```
No particular reason, it just seemed like 10 would be enough for this particular use case. Personally it seems extremely unlikely that someone would want to filter out based on more than 10 different labels. 

What do you think?

itsdalmo replied:
> Personally it seems extremely unlikely that someone would want to filter out based on more than 10 different labels.

But this is the number of labels to retrieve _before_ you start filtering? So if you have e.g. 11 labels on a PR, then the one you are looking for might not be included in the response (and the PR would be filtered out by mistake)?

Is there an upper limit for how many labels a PR can have? 🤔

sergiughf replied:
> But this is the number of labels to retrieve before you start filtering? So if you have e.g. 11 labels on a PR, then the one you are looking for might not be included in the response (and the PR would be filtered out by mistake)?

Yep that's correct, but from my experience we've always used labels for a specific purpose and we've hardly used more than 3-4 labels at a time and the labels would always have some kind of filtering purpose. But obviously this is just my use case:) What would be a good value in your opinion? Try to aim really high at around 100 maybe?

I tried looking a bit around but can't seem to find a limit to how many labels you could have on a PR on the official github docs. https://help.github.com/en/articles/applying-labels-to-issues-and-pull-requests I sent a help request to github support to clear this up for us. :) Will let you know what they answer.

Meanwhile maybe we just set it to 100 then?

sergiughf commented:
Hey @itsdalmo thanks for reviewing the PR! Took care of addressing your comments, could you please take another look whenever you have time?

The only questions I would still have are:

> Otherwise I feel like I/we need to review how costs for the API calls are calculated again to make sure that fetching labels as well will not affect peoples usage toward their rate limits (if that is the case then I think we need to make labels opt-in).
- How exactly could we achieve this? Could you give some pointers?

- How can we add e2e tests for the extra label logic on the `check` binary?

sergiughf commented:
Hey @itsdalmo I still didn't hear anything from github support yet. But would you like me to do any further changes on this PR?:)

itsdalmo commented:
@sergiughf I'm sorry I have not had a chance to look at the changes yet :/ Let me take a look during the weekend if that's alright?

sergiughf commented:
Yep that's perfectly fine no worries!:)

sergiughf commented:
@itsdalmo any updates?:)

itsdalmo commented on in.go:
```diff
@@ -63,7 +63,7 @@ func Get(request GetRequest, github Github, git Git, outputDir string) (*GetResp
 
 	// Create the metadata
-	var metadata Metadata
+	metadata := Metadata{}
```
What is the reason for changing this? 😅

sergiughf replied:
To remove the warning `Goland` throws here :) for `Receiver 'metadata' may be nil in call`.

itsdalmo commented on in_test.go:
```diff
@@ -40,7 +40,7 @@ func TestGet(t *testing.T) {
 			},
 			parameters:     resource.GetParameters{},
-			pullRequest:    createTestPR(1, "master", false, false),
+			pullRequest:    createTestPR(1, "master", false, false, []string{"help wanted"}),
```
I think that for `in_test.go` we can just pass a `nil` as the labels for the fake pull requests.

sergiughf commented:
I'll try to spend some time the following days to also rebase and fix everything :+1:

sergiughf commented:
Hey @itsdalmo :) Do you mind taking a look again? :pray: 

I've updated the PR, rebased and cleaned the git commit history.

sergiughf commented:
Thank you for merging this in @itsdalmo ! yay!

Came back to update you on the github inquiry on label limit I had opened. After almost a month github support finally decided to answer:) 

```
Aug 20, 9:33 AM UTC

Hi Sergiu,

My apologies for the time it's taken to respond! We really do appreciate your patience.

There's a maximum limit of 100 labels for a Pull Request.

Cheers
Stacey
```
